### DISCUSSION TITLE
How to keep trying until establishing ethernet connection

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#9853] [Using MicroPython] [answered]
Hi,

I'm programming a wiznet w5100s evb Pico (Pico with integrated ethernet connector). 
I would like it to keep trying to connect indefinitely, until a connection is made.

At the moment I have the code below which works, sort of:

-- The init function runs
-- On a connection error the function 'runs itself again'
-- The code makes 39 attempts
-- This takes about 10 minutes
-- If there is still no connection I get: 'RuntimeError: maximum recursion depth exceeded'

What could be a different approach to keep trying indefinitely?
In Python I used threading in a similar situation but that's not available in MicroPython right?

EDIT:
I just found that I might be able to use _thread since the Pico has two cores.
Any suggestions on how to achieve this would still be very welcome though
/EDIT

Thanks!

My connecting function:
```
connection_attempts = 0

def init_ethernet():

    try:
        # Keep track of number of attempts
        global connection_attempts
        connection_attempts += 1
        print(f'Connecting... attempt: {connection_attempts}')
        # Try nic.active takes 17-18 seconds
        nic.active(True)                                        
        
    except OSError as error:
        now = time.localtime()
        print(f'Connection error: {error}, {now[3]:02d}:{now[4]:02d}:{now[5]:02d}')
        # In case of connection error, run this function again
        init_ethernet()

    ip_address = nic.ifconfig()[0]
    subnet = nic.ifconfig()[1]
    gateway = nic.ifconfig()[2]
    dns = nic.ifconfig()[3]

    print('Connected:')
    print('IP        ', ip_address)
    print('Subnet    ', subnet)
    print('Gateway   ', gateway)
    print('DNS       ', dns)

    # create_socket()
    server_socket.bind((ip_address, socket_port))
    server_socket.listen(1)

    accept_socketclient()
```

The output:
```
Connecting... attempt: 1
Connection error: timeout waiting for DHCP to get IP address, 09:14:13
...
...
Connection error: timeout waiting for DHCP to get IP address, 09:24:42
Connecting... attempt: 39
Traceback (most recent call last):
  File "<stdin>", line 149, in <module>
....
  File "<stdin>", line 54, in init_ethernet
RuntimeError: maximum recursion depth exceeded
```

**Comment by karfas** [comment, marked as answer]
You are calling init_ethernet() from inside itself, hence the recursion.
A simple loop will use way less resources, e..g. 

```python
have_network = False
while not have_network:
    try:
        # Keep track of number of attempts
        global connection_attempts
        connection_attempts += 1
        print(f'Connecting... attempt: {connection_attempts}')
        # Try nic.active takes 17-18 seconds
        nic.active(True)                                        
        have_network = True
        
    except OSError as error:
        now = time.localtime()
        print(f'Connection error: {error}, {now[3]:02d}:{now[4]:02d}:{now[5]:02d}')
        # In case of connection error, run this function again
```

**Reply by MrTinkerman** [reply]
Thanks karfas, it works :)

The short version:
```
    connected = False
    
    while not connected:
        try:
            nic.active(True)
            connected = True
        except:
            pass
```

**Comment by robert-hh** [comment]
Besides to what @karfas said, you could set up a timer with a callback, which tries to connect, and when successful, either stops the timer or continues, watching the connections and reconnects, if the connection was broken. If you do that within a class, you can have the status information bound to the class.

**Reply by MrTinkerman** [reply]
Thanks for the suggestion, it would be a good idea to monitor the connection and reconnect if needed.
Next goal has been set :)
